order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java

```diff
@@ -1,58 +1,179 @@
+import com.orderbook.OrderBookLayer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.TreeSet;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 @Component
 public class ItemMarket {
 
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
     private final Map<String, TreeSet<Item>> allItems;
+    private final Map<String, Long> productsQuantity;
+
+    private static final Logger LOGGER = LogManager.getLogger(ItemMarket.class);
 
     @Autowired
     public ItemMarket() {
-        allItems = new ConcurrentHashMap<>();
+        this.allItems = new ConcurrentHashMap<>();
+        this.productsQuantity = new ConcurrentHashMap<>();
         addDummyData();
     }
 
     public Optional<Set<Item>> getItemSetByName(String itemName) {
-        return Optional.of(allItems.get(itemName));
+        return Optional.of(this.allItems.get(itemName));
     }
 
     public Set<String> getItemNameSet() {
-        return allItems.keySet();
+        return this.allItems.keySet();
     }
 
     public void addTrackedItem(String itemName) {
-        allItems.put(itemName, new TreeSet<>());
+        this.allItems.put(itemName, new TreeSet<>());
     }
 
     public void removeUntrackedItem(String itemName) {
-        allItems.remove(itemName);
+        this.allItems.remove(itemName);
+    }
+
+    public void updateItem(Aurora.AuroraResponse response) {
+
+        if (!response.getMessage().is(TickResponse.class)) {
+            throw new IncorrectResponseException("Response is not correct!");
+        }
+
+        TickResponse tickResponse;
+
+        try {
+            tickResponse = response.getMessage().unpack(TickResponse.class);
+        } catch (InvalidProtocolBufferException e) {
+            throw new IncorrectResponseException("Response is not correct!");
+        }
+        Set<Item> itemSet = this.allItems.get(tickResponse.getGoodName());
+        if (itemSet == null) {
+            LOGGER.error("Item: {} is not being tracked and cannot be added to itemMarket!",
+                    tickResponse.getGoodName());
+            return;
+        }
+        Item item = populateItem(tickResponse);
+
+        this.productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+        LOGGER.info("Products data updated!");
+
+        if (itemSet.contains(item)) {
+
+            Item presentItem = itemSet.stream().filter(currentItem -> Double.compare(currentItem.getPrice(), item.getPrice()) == 0
+                    && currentItem.getOrigin().equals(item.getOrigin())).findFirst().get();
```

Comment:
please use item.compareTo, as the functionality is the same

Reply:
Fixed